recipes/CoVoST/AST/train_w2v2_llama.py

```diff
@@ -0,0 +1,550 @@
+
+        # LLM forward
+        llm_logits = self.modules.lora_llm(
+            inputs_embeds=inputs_embeds, attention_mask=attn_mask
+        ).logits
+
+        # output layer for seq2seq log-probabilities
+        p_seq = self.hparams.log_softmax(llm_logits)
+
+        if hasattr(self.modules.lora_llm, "module"):
+            gen_func = self.modules.lora_llm.module.model.generate
+        else:
+            gen_func = self.modules.lora_llm.model.generate
+
+        # Running decoding if not training
+        if stage == sb.Stage.TRAIN:
+            hyps = None
+
+        elif stage == sb.Stage.VALID:
+
+            # Define generation config depending on runtime values
+            config = transformers.GenerationConfig(
+                num_beams=self.hparams.valid_beam_size,
+                pad_token_id=tokenizer.pad_token_id,
+                eos_token_id=tokenizer.eos_token_id,
+                max_new_tokens=int(2 * audio_len),
+            )
+
+            hyps = gen_func(
+                inputs_embeds=inputs_embeds[
+                    :, :audio_prompt_len
+                ],  # give model audio features and prompt for inference
+                attention_mask=attn_mask[:, :audio_prompt_len],
+                generation_config=config,
+            )
+        elif stage == sb.Stage.TEST:
+
+            # Define generation config depending on runtime values
+            config = transformers.GenerationConfig(
+                num_beams=self.hparams.test_beam_size,
+                pad_token_id=tokenizer.pad_token_id,
+                eos_token_id=tokenizer.eos_token_id,
+                max_new_tokens=int(2 * audio_len),
+            )
+
+            hyps = gen_func(
+                inputs_embeds=inputs_embeds[
+                    :, :audio_prompt_len
+                ],  # give model audio features and prompt for inference
+                attention_mask=attn_mask[:, :audio_prompt_len],
+                generation_config=config,
+            )
+
+        return p_seq, wav_lens, hyps, audio_prompt_len
+
+    def compute_objectives(self, predictions, batch, stage):
+        """Computes the loss (CTC+NLL) given predictions and targets."""
+
+        (
+            p_seq,
+            wav_lens,
+            predicted_tokens,
+            audio_prompt_len,
+        ) = predictions
+
+        ids = batch.id
+        tokens_translation, tokens_translation_len = batch.tokens_translation
+
+        # Translation loss
+        # We are only interested in computing the loss over the logits after
+        # the audio + prompt embeddings. Tokens_translation does not start with bos, so we just need to make sure to shift the logits to the last token of the prompt (to ensure next word prediction)
+
+        p_seq_translation_only = p_seq[:, audio_prompt_len - 1 :]
+
+        loss = self.hparams.nll_loss(
+            p_seq_translation_only,
+            tokens_translation,
+            length=tokens_translation_len,
+        )
```

Comment:
I don't know what is more beautiful. Setting the target labels for `audio embds + prompt embeddings` to be `-100` or just cropping the input + target?